Do not apply environment maps to unlit models. Fixes #374

jsantell opened the pull request:
Currently with a default env map, image env map, and PMREM'd env map currently:

![localhost_5000_examples_fullscreen2 html 11](https://user-images.githubusercontent.com/641267/53048345-60a20a80-3449-11e9-9a91-837f523ec2b5.png)

With this patch:

![localhost_5000_examples_fullscreen2 html 10](https://user-images.githubusercontent.com/641267/53048371-6e579000-3449-11e9-979e-f7aebcb34719.png)

Note that `UnlitTest.glb` is included (4kb) -- can reference the Khronos models if/when they land. The Astronaut is *not* unlit AFAICT (or, does not use the unlit extension), maybe should change the background-image demo describing it as such.

donmccurdy commented on src/three-components/Model.js:
```diff
@@ -50,7 +50,10 @@ export default class Model extends Object3D {
+      // Unlit models (using MeshBasicMaterial) should not apply
+      // an environment map, even though `map` is the currently configured
+      // environment map.
+      if (obj && obj.isMesh && obj.material && !obj.material.isMeshBasicMaterial) {
```
For some edge cases, `obj.material` could be an array of materials. We should probably avoid creating those in GLTFLoader at some point, but for now it's worth handling.

cdata commented:
@jsantell since we have merged #375 , would you be willing to update this change to use the unlit sample from the Khronos samples so that we don't have to check the model into our repo?

jsantell commented:
@cdata  
> since we have merged #375 , would you be willing to update this change to use the unlit sample from the Khronos samples so that we don't have to check the model into our repo?

wouldn't that require needing to fetch all the models before being able to run tests? UnlitTest.glb is 4kb as well.

@donmccurdy 
> For some edge cases, `obj.material` could be an array of materials. We should probably avoid creating those in GLTFLoader at some point, but for now it's worth handling.

Any models that you know that do this to test?

cdata commented:
>> since we have merged #375 , would you be willing to update this change to use the unlit sample from the Khronos samples so that we don't have to check the model into our repo?
> wouldn't that require needing to fetch all the models before being able to run tests? UnlitTest.glb is 4kb as well.

Yes, and we do this (models are fetched as part of `npm prepare`, so they happen after `npm install` and `npm ci`).

In practice this is probably fine. We have set up Travis to cache the directory they are checked out to, and it doesn't seem to take much time. We also update in place if we detect that the models are already fetched.

cdata commented:
FYI looks like `UnlitTest.glb` is still checked in

jsantell commented:
Using the Khronos sample models now, and handle when `material` is an array

cdata commented on src/three-components/Model.js:
```diff
@@ -49,10 +49,24 @@ export default class Model extends Object3D {
+        materialStack.push(obj.material);
+      }
+
+      for (let material of materialStack) {
```
Did you mean to do iteration of the material stack on every invocation of the `traverse` callback?

jsantell replied:
Mm, yes, but missing clearing it out. I don't like any way that attempts to not duplicate the logic for the (untested) edge case as well. Will change

donmccurdy commented:
It's a bit contrived, but here's an example that will result in `mesh.material` being an array. The model contains a single mesh with two primitives, and the primitives are identical except for using different materials.

[Triangle.gltf.zip](https://github.com/GoogleWebComponents/model-viewer/files/2902385/Triangle.gltf.zip)

jsantell commented:
* Fixed the material array logic
* Added Triangle.gltf for test case (added PBR values to see envmap), shook out some issues in our GLTFLoader wrapper
* Remove UnlitTest.glb

cdata commented on src/test/features/environment-spec.js:
```diff
@@ -160,6 +175,34 @@ suite('ModelViewerElementBase with EnvironmentMixin', () => {
+        });
+      });
+
+      suite('on a model with multi-material meshes', () => {
```
💯
